src/aframe_components/custom-components.ts

```diff
@@ -20,24 +25,100 @@ interface State {
 export interface Guidance extends Component {
   state: State;
   pointPositions: Vector3[];
+  kdtree: KDNode | null;
+}
+
+export interface KDNode {
+  position: Vector3;
+  left: KDNode | null;
+  right: KDNode | null;
+}
+
+export function build3DTree(positions: Vector3[], depth: number){
+  const n = positions.length;
+  const mid = Math.floor(n / 2);
+  if (n <= 0){
+    return null;
+  }
+  const axis = depth % 3;
+  positions.sort((a, b) => a[AXIS_DICT[axis]] - b[AXIS_DICT[axis]]);
+  return {
+    position: positions[mid],
+    left: build3DTree(positions.slice(0, mid), depth + 1),
+    right: build3DTree(positions.slice(mid + 1, n), depth + 1),
+  };
+}
+
+export function nearestPoint(controllerPos: Vector3, node: KDNode | null, depth: number){
+  if (node == null){
+    return null;
+  }
+
+  const axis = depth % 3;
+  let nextNode: KDNode | null = null;
+  let otherNode: KDNode | null = null;
+
+  if (controllerPos[AXIS_DICT[axis]] < node.position[AXIS_DICT[axis]]){
+    nextNode = node.left;
+    otherNode = node.right;
+  }
+  else{
+    nextNode = node.right;
+    otherNode = node.left;
+  }
+
+  let closest = closerDistance(
+    controllerPos,
+    node.position,
+    nearestPoint(controllerPos, nextNode, depth + 1));
+
+  if (controllerPos.distanceTo(closest!) > Math.abs(controllerPos[AXIS_DICT[axis]] - node.position[AXIS_DICT[axis]])){
+    closest = closerDistance(
+      controllerPos,
+      closest,
+      nearestPoint(controllerPos, otherNode, depth + 1));
+  }
+
+  return closest;
+}
+
+function closerDistance(controllerPos: Vector3, pointA: Vector3 | null, pointB: Vector3 | null){
+  if (pointA == null){
+    return pointB;
+  }
+  if (pointB == null){
+    return pointA;
+  }
+
+  const distanceA: number = controllerPos.distanceTo(pointA);
+  const distanceB: number = controllerPos.distanceTo(pointB);
+
+  if (distanceA < distanceB){
+    return pointA;
+  }
+  else{
+    return pointB;
+  }
```

Comment:
Probably a good idea to move the tree node implementation into its own file.  Then it can be tested independently, and can also be used elsewhere (theoretically)